src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,256 @@
+        return;
+      bool is_read = is_read_op(method);
+      if ((is_read && !info->max_read_bytes) || (!is_read && !info->max_write_bytes))
+        return;
+      auto& it = find_or_create(key);
+      std::unique_lock lock(it.ts_lock);
+      // we don't want the tenant to be with higher debt than 120 seconds(2 min) of its limit
+      if (is_read)
+        it.read.bw = std::max(it.read.bw - amount * fixed_float_rgw_ratelimit,info->max_read_bytes * fixed_float_rgw_ratelimit * -2);
+      else {
+        it.write.bw = std::max(it.write.bw - amount * fixed_float_rgw_ratelimit,info->max_write_bytes * fixed_float_rgw_ratelimit * -2);
+      }
+      ldpp_dout(dpp, 20) << "rate limiting read bytes is: " << it.read.bw <<
+                  " write bytes is: " <<  it.write.bw << dendl;
+      ldpp_dout(dpp, 20) << "rate limiting key is: " << key << " bytes added: " << amount <<
+      " method is: " << method << dendl;
+    }
+    void clear() {
+      qos_entries.clear();
+    }
+};
+// This class purpose is to hold 2 RateLimiter instances, one active and one passive.
+// once the active has reached the watermark for clearing it will call the replace_active() thread using cv
+// The replace_active will clear the previous RateLimiter after all requests to it has been done (use_count() > 1)
+// In the meanwhile new requests will come into the newer active
+class ActiveRateLimiter : public DoutPrefix  {
+  std::atomic_uint8_t stopped = {false};
+  std::condition_variable cv;
+  std::mutex cv_m;
+  std::thread runner;
+  std::atomic_bool replacing = false;
+  std::atomic_uint8_t current_active = 0;
+  std::shared_ptr<RateLimiter> qos[2];
+  void replace_active() {
+    using namespace std::chrono_literals;
+    std::unique_lock<std::mutex> lk(cv_m);
+    while(!stopped) {
+      cv.wait(lk);
+      current_active = current_active ^ 1;
+      ldpp_dout(this, 20) << "replacing active qos data structure" << dendl;
+      while (!stopped && qos[(current_active ^ 1)].use_count() > 1 ) {
+        if (stopped && cv.wait_for(lk, 1min) != std::cv_status::timeout)
+          return;
+      }
+      if (stopped)
+        return;
+      ldpp_dout(this, 20) << "clearing passive qos data structure" << dendl;
+      qos[(current_active ^ 1)]->clear();
+      replacing = false;
+    }
+  }
+  public:
+    ActiveRateLimiter(const ActiveRateLimiter&) = delete;
+    ActiveRateLimiter& operator =(const ActiveRateLimiter&) = delete;
+    ActiveRateLimiter(ActiveRateLimiter&&) = delete;
+    ActiveRateLimiter& operator =(ActiveRateLimiter&&) = delete;
+    ActiveRateLimiter() = delete;
+    ActiveRateLimiter(CephContext* cct) : 
+      DoutPrefix(cct, ceph_subsys_rgw, "rate limiter: ")
+    {
+      qos[0] = std::make_shared<RateLimiter>(&replacing, &cv);
+      qos[1] = std::make_shared<RateLimiter>(&replacing, &cv);
+    }
+    ~ActiveRateLimiter() {
+      ldpp_dout(this, 20) << "stopping ratelimit_gc thread" << dendl;
+      cv_m.lock();
```

Comment:
Without locking this variable there are situations that the notify will arrive before the wait and the thread will stuck on wait()

Reply:
to avoid those races between notify and wait, the mutex should be locked over the call to `notify_all()`. `stopped = true` doesn't necessarily need to happen inside the lock, but it might as well be